python/ray/autoscaler/_private/autoscaler.py

```diff
@@ -50,15 +52,55 @@
 
 logger = logging.getLogger(__name__)
 
+# Status of a node e.g. "up-to-date", see ray/autoscaler/tags.py
+NodeStatus = str
+
 # Tuple of modified fields for the given node_id returned by should_update
 # that will be passed into a NodeUpdaterThread.
 UpdateInstructions = namedtuple(
     "UpdateInstructions",
     ["node_id", "setup_commands", "ray_start_commands", "docker_config"])
 
-AutoscalerSummary = namedtuple(
-    "AutoscalerSummary",
-    ["active_nodes", "pending_nodes", "pending_launches", "failed_nodes"])
+
+@dataclass
+class AutoscalerSummary:
+    active_nodes: Dict[NodeType, int]
+    pending_nodes: List[Tuple[NodeIP, NodeType, NodeStatus]]
+    pending_launches: Dict[NodeType, int]
+    failed_nodes: List[Tuple[NodeIP, NodeType]]
+
+
+class NonTerminatedNodes:
+    """Class to extract and organize information on non-terminated nodes."""
+
+    def __init__(self, provider: NodeProvider):
+        # All non-terminated nodes
+        self.all_nodes = provider.non_terminated_nodes({})
+
+        # Managed worker nodes (node kind "worker"):
+        self.workers: List[NodeID] = []
+        # The head node (node kind "head")
+        self.head: Optional[NodeID] = None
+
+        for node in self.all_nodes:
+            node_kind = provider.node_tags(node)[TAG_RAY_NODE_KIND]
+            if node_kind == NODE_KIND_WORKER:
+                self.workers.append(node)
+            elif node_kind == NODE_KIND_HEAD:
+                self.head = node
+
+        # Note: For typical use-cases,
+        # self.all_nodes == self.workers + self.head
```

Comment:
Turns out it's better to sweep unmanaged workers under the rug and to have an independent list of all_nodes.
Works better with the tests and ReadOnlyNodeProvider.